src/main/kotlin/com/valbaca/gotlin/ch8/Thumbnail.kt

```diff
@@ -64,109 +67,55 @@ suspend fun makeThumbnails2(filenames: List<String>) = coroutineScope {
-    }
-    repeat(filenames.size) {
-        val e = exceptions.receive()
-        if (e != null) {
-            throw e // Unlike go, this will NOT result in a coroutine leak. Per the documentation of `coroutineScope`
-            // However, if we used supervisorScope, we would need to protect against this. See makeThumbnails5
-            /* When any child coroutine in this scope fails, this scope fails and all the rest of the children are
-             cancelled (for a different behavior see supervisorScope). */
-        }
-    }
+suspend fun makeThumbnails4(filenames: List<String>) = supervisorScope {
+    filenames.forEach { launch { imageFile(it) } }
 }
 
 /*
 https://github.com/adonovan/gopl.io/blob/master/ch8/thumbnail/thumbnail_test.go#L86
- */
-suspend fun makeThumbnails5(filenames: List<String>): List<String> = supervisorScope {
-    // TIP: Consider Arrow-kt's Either https://apidocs.arrow-kt.io/arrow-core/arrow.core/-either/index.html
-    data class Item(val thumbfile: String?, val error: Exception?)
-
 
-    val ch = Channel<Item>(filenames.size)
-    for (f in filenames) {
-        launch {
-            // This imitates Go's multi-return functionality
-            val item = try {
-                Item(imageFile(f), null)
-            } catch (e: Exception) {
-                Item(null, e)
-            }
-            ch.send(item)
-        }
-    }
-    buildList {
-        repeat(filenames.size) {
-            val item = ch.receive()
-            if (item.error != null) {
-                throw item.error
-            }
-            add(item.thumbfile!!)
-        }
-    }
+Unlike `launch`, which is a fire-and-forget job, `async` will return the result of the imageFile
+method. The `awaitAll` method will wait for all the jobs to complete and return a list of
+the results. The first async job that throws an exception will cause any other jobs in progress
+to be cancelled and the exception will be propagated.
+ */
+suspend fun makeThumbnails5(filenames: List<String>): List<String> = coroutineScope {
+    filenames.map { async { imageFile(it) } }.awaitAll()
 }
 
 /*
 https://github.com/adonovan/gopl.io/blob/master/ch8/thumbnail/thumbnail_test.go#L117
  */
 suspend fun makeThumbnails6(filenames: ReceiveChannel<String>): Long = coroutineScope {
-    val sizes = Channel<Long>()
-    val jobs = mutableListOf<Job>()
-    // I *think* this is the Kotlin equivalent of a sync.WaitGroup. Please correct or confirm!
-    // https://github.com/adonovan/gopl.io/blob/master/ch8/thumbnail/thumbnail_test.go#L119
-    for (f in filenames) {
-        val job = launch {
-            try {
-                val thumb = imageFile(f)
-                sizes.send(stat(thumb))
-            } catch (e: Exception) {
-                throw e
-            }
-        }
-        jobs.add(job)
-    }
-    launch {
-        for (job in jobs) {
-            job.join() // join on the jobs rather than use a WaitGroup
+    buildList {
+        filenames.consumeEach {
```

Comment:
`consumeEach` is the same as `for (x in channel)` with the addition that it cancels the channel when it exits.